frost/participant.go

```diff
@@ -0,0 +1,424 @@
+	//   main protocol.  A polynomial of maximum degree t is represented as a
+	//   list of t+1 coefficients, where the constant term of the polynomial
+	//   is in the first position and the highest-degree coefficient is in the
+	//   last position.  For example, the polynomial x^2 + 2x + 3 has degree 2
+	//   and is represented as a list of 3 coefficients [3, 2, 1].  A point on
+	//   the polynomial f is a tuple (x, y), where y = f(x).
+	//
+	//   The function derive_interpolating_value derives a value used for
+	//   polynomial interpolation.  It is provided a list of x-coordinates as
+	//   input, each of which cannot equal 0.
+	//
+	//   Inputs:
+	//     - L, the list of x-coordinates, each a NonZeroScalar.
+	//     - x_i, an x-coordinate contained in L, a NonZeroScalar.
+	//
+	//   Outputs:
+	//     - value, a Scalar.
+	//
+	//   Errors:
+	//     - "invalid parameters", if 1) x_i is not in L, or if 2) any
+	//       x-coordinate is represented more than once in L.
+	//
+	//   def derive_interpolating_value(L, x_i):
+
+	order := p.ciphersuite.Curve().Order()
+	// numerator = Scalar(1)
+	num := big.NewInt(1)
+	// denominator = Scalar(1)
+	den := big.NewInt(1)
+	// for x_j in L:
+	for _, xj := range L {
+		if xj == xi {
+			// if x_j == x_i: continue
+			continue
+		}
+		// numerator *= x_j
+		num.Mul(num, big.NewInt(int64(xj)))
+		num.Mod(num, order)
+		// denominator *= x_j - x_i
+		den.Mul(den, big.NewInt(int64(xj)-int64(xi)))
+		den.Mod(den, order)
+	}
+
+	// value = numerator / denominator
+	denInv := new(big.Int).ModInverse(den, order)
+	res := new(big.Int).Mul(num, denInv)
+	res = res.Mod(res, order)
+
+	// return value
+	return res
+}
+
+// computeChallenge implements def compute_group_commitment(commitment_list,
+// binding_factor_list) from [FROST] as defined in section 4.6. Signature
+// Challenge Computation.
+func (p *Participant) computeChallenge(
+	message []byte,
+	groupCommitment *Point,
+) *big.Int {
+
+	// From [FROST]:
+	//
+	// 4.6.  Signature Challenge Computation
+	//
+	//   This section describes the subroutine for creating the per-message
+	//   challenge.
+	//
+	//   Inputs:
+	//     - group_commitment, the group commitment, an Element.
+	//     - group_public_key, the public key corresponding to the group signing
+	//       key, an Element.
+	//     - msg, the message to be signed, a byte string.
+	//
+	//   Outputs:
+	//     - challenge, a Scalar.
+	//
+	// def compute_group_commitment(commitment_list, binding_factor_list)
+
+	curve := p.ciphersuite.Curve()
+	// group_comm_enc = G.SerializeElement(group_commitment)
+	groupCommitmentEncoded := curve.SerializePoint(groupCommitment)
+	// group_public_key_enc = G.SerializeElement(group_public_key)
+	publicKeyEncoded := curve.SerializePoint(p.publicKey)
+	// challenge_input = group_comm_enc || group_public_key_enc || msg
+	// challenge = H2(challenge_input)
+	// return challenge
+	return p.ciphersuite.H2(groupCommitmentEncoded, publicKeyEncoded, message)
```

Comment:
We need to consider the difference between the internal encoding and the external encoding here, and this will need to use the external encoding instead of `SerializePoint`. For example, in BIP-340 the external encoding is not lossless.

I see two options for this: we can either add something like `EncodePoint` and `DecodePoint` to the `Curve` interface, or we can pass `*Point` arguments to `H2`. The latter is a bit more robust in case someone comes up with an esoteric protocol that wants something weird done with the points, but the former may be a bit more convenient for encoding the signatures into `[]byte` from the `(*Point, *big.Int)` they are produced as, because it's likely that the same functions would be used in both the signature encoding and the challenge computation argument encoding.

Reply:
As far as I can tell, this is only a problem for H2 which is used to compute the challenge, correct?

In that case, I'd probably go with the more robust version, which is passing `*Point` arguments to `H2`. *If* `H2` is the only function that requires special attention in this matter, I think it is fine if we make it clear in the function signature. Moreover, we already have `SerializePoint` and `DeserializePoint` in the public API of the curve and I think having serialize and encode next to each other could be confusing. 

Alternatively, and I think I like this version most, we could expose `EncodePoint` / `DecodePoint` on the `Hashing` or `Ciphersuite` interfaces.

Reply:
I think you're right with that; I'm not aware of anywhere apart from H2 which needs to use the external encoding.

I like the idea of exposing the other function on a different interface to make it clear that it's related to how the signature scheme expects things to be formatted and separate the encodings conceptually.

Making `H2` take `*Point` arguments _and_ exposing the external encoding on the `Ciphersuite` interface would be the _most_ robust just in case someone has a different encoding for challenge computation vs. signatures (code-wise it would be more elegant as part of `Hashing` but conceptually fits better in `Ciphersuite` in this case). However, that feels like potential YAGNI.

Reply:
Please see the proposition from f6af9a5a7dae5f7c736581a176cc65bd8cb0b6c9.

Reply:
Thinking more about it, I would say `EncodePoint` function fits better `Hashing` interface but it is a cosmetic change I would do in another PR to do not complicate merges (#10 is branched off already).